Feature/SAAS-2987-reconnect-when-rabbit-connection-is-closed

rafiramadhana opened the pull request:
Hi, this is Rafi.

This PR will implement reconnect when our connection to RabbitMQ is closed by server.

Jira:
[SAAS-2987](https://shortlyst.atlassian.net/browse/SAAS-2987)

Changes:
- Add `RabbitMqClient.url` to store url of given RabbitMQ host
- Add `RabbitMqClient.rcWg` as wait group for reconnect process
- Add `RabbitMqClient.rcStepTime` as time between reconnect attempts
- Add `RabbitMqClient.maxRc` as maximum number of attempts to reconnect
- Implement `amqp.Channel.NotifyClose` to listen when server sends a channel or connection exception 
- Add `RabbitMqClient.reconnect()` as reconnect implementation
- Add `RabbitMqClient.SetRcStepTime()` as setter for `.rcStepTime` field

Notes:
- This PR only handles reconnecting the RabbitMQ connection and reconsume the subscribers (if any). The republishing process will be done in another PR.
- I am still figuring out how to make integration and/or unit tests for this new implementation, any help from you guys might be helpful. 
- So far, what I did was testing this new implementation by manually dropping the connection from the RabbitMQ management interface. 

References:
- [Documentation of amqp.Channel.NotifyClose](https://pkg.go.dev/github.com/streadway/amqp#Channel.NotifyClose)

p.s. I don't know what is Bob's and Ryan's github account, so maybe we can mention to them later about this PR. Thank you.

rf-shortlyst committed:
Add reconnect implementation

rf-shortlyst committed:
Expose setter for rcStepTime

hendra-shortlyst commented on rabbitmq.go:
```diff
@@ -1,27 +1,43 @@
 }
 
-func NewRabbitMqClient(conn string, threadNum int) *RabbitMqClient {
+func NewRabbitMqClient(conn string, threadNum int, maxReconnect int) *RabbitMqClient {
```
@neverbeenthisweeb I think it would be better to introduce minimal changes to the user of this library by moving the maxReconnect to Setter function and set reasonable default value if not set. Wdyt?

rafiramadhana replied:
I see, this is a good idea. For the default, maybe we could follow AWS SDK's default retry `max_attempts` (see [here](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/retries.html#standard-retry-mode)), which is `3`. Wdyt?

hendra-shortlyst replied:
@neverbeenthisweeb since this lib usually used in long running application, I think it would be better to make reconnect attempt to be indefinite by default. If we choose this, the question would be: will there be a significant drawbacks on reconnecting indefinitely?

rafiramadhana replied:
Ok, since the step time between reconnect attempts is increasing gradually (see [here](https://github.com/shortlyst-ai/go-msgbuzz/pull/2/files#diff-c8d676eb0bcb860e00da09512c95b2f0R221)), I think there is no major drawbacks. Wdyt?

rf-shortlyst committed:
Remove maxRc

This will implement indefinite reconnect attempt to RabbitMQ connection

rf-shortlyst committed:
Rename i to currentRcAttempt

rf-shortlyst committed:
Remove redundant log on reconnect

This log will only be reached  when StartConsuming is ended

It happens because StartConsuming is a blocking call
